fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,120 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    this.getAllTodos();
+  }
+
+  getAllTodos() {
+    fetch('api/todos')
+      .then(res => res.json())
+      .then(data => this.setState({ todos: data })
+      );
+  }
+  /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+
+  addTodo(newTodo) {
+    fetch('api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(res => res.json())
+      .then(newEntry => {
+        const updatedToDo = [...this.state.todos];
+        updatedToDo.push(newEntry);
+        this.setState({ todos: updatedToDo });
+      });
+  }
+  /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * Do not mutate the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+
+  toggleCompleted(todoId) {
+    let matchingToDo;
+    for (let i = 0; i < this.state.todos.length; i++) {
+      const currentToDo = this.state.todos[i];
+      if (currentToDo.todoId === todoId) {
+        matchingToDo = i;
+      }
+    }
+    const todos = this.state.todos;
```

Comment:
You want to make sure to copy the `todos` array, this only creates a pointer to the original array. You can do one of the following:

```JAVASCRIPT
const todos = [...this.state.todos];

// --- OR ---

const todos = this.state.todos.slice();
```